distributed/worker_state_machine.py

```diff
@@ -3722,6 +3722,134 @@ def digest_metric(self, name: str, value: float) -> None:
+        property is read.
+
+        Parameters
+        ----------
+        by_prefix: bool, optional
+            True (default)
+                Return mapping of (task prefix, task state) -> seconds
+            False
+                Return mapping of task state -> seconds
+        """
+        # This would prevent generating a cluster dump when the state is corrupted
+        # assert not self._new_tasks, self._new_tasks
+
+        if self._previous_ts is None:
+            assert not self._current_count
+            assert not self._cumulative_elapsed
+        else:
+            now = monotonic()
+            elapsed = now - self._previous_ts
+            self._previous_ts = now
+            for k, n_tasks in self._current_count.items():
+                self._cumulative_elapsed[k] += elapsed * n_tasks
+
+        if by_prefix:
+            return self._cumulative_elapsed
+
+        out: defaultdict[TaskStateState, float] = defaultdict(float)
+        for (_, state), n in self._cumulative_elapsed.items():
+            out[state] += n
+        return out
+
+    def new_task(self, ts: TaskState) -> None:
+        """A new task has just been created and will be immediately fed into the
+        recommendations for transitions
+        """
+        self._new_tasks.add(ts)
+
+    def transitions(self, prev_states: dict[TaskState, TaskStateState]) -> None:
+        """Tasks have just transitioned to a new state"""
+        if not prev_states and not self._new_tasks:
+            return
+
+        now = monotonic()
+        if self._previous_ts is not None:
+            elapsed = now - self._previous_ts
+
+        for ts, prev_state in prev_states.items():
+            if ts in self._new_tasks:
+                self._new_tasks.discard(ts)
+                continue
+            if ts.state == prev_state:
+                continue
+
+            k = ts.prefix, prev_state
+            if self._previous_ts is not None:
+                self._cumulative_elapsed[k] += elapsed
+
+            new_count = self._current_count[k] - 1
+            if new_count > 0:
+                self._current_count[k] = new_count
+            else:
+                assert new_count == 0
+                del self._current_count[k]
+
+        if self._previous_ts is not None:
+            for k, n_tasks in self._current_count.items():
+                self._cumulative_elapsed[k] += elapsed * n_tasks
+
+        for ts, prev_state in prev_states.items():
+            if ts.state not in (prev_state, "forgotten"):
+                self._current_count[ts.prefix, ts.state] += 1
+
+        for ts in self._new_tasks:
+            # This happens exclusively on a transition from cancelled(flight) to
+            # resumed(flight->waiting) of a task with dependencies; the dependencies
+            # will remain in released state and never transition to anything else.
+            self._current_count[ts.prefix, ts.state] += 1
```

Comment:
I find quite scary that this code path is tripped by exactly one test in the whole test suite (`test_deadlock_cancelled_after_inflight_before_gather_from_worker`)